kdc/bx509d.c

```diff
@@ -1843,6 +1939,189 @@ get_tgt(struct bx509_request_desc *r)
+            (void) close(fd);
+            return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                           "%s", strerror(r->error_code = errno));
+        }
+    }
+    fn = strchr(r->ccname, ':');
+    if (fn == NULL)
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "Internal error (invalid credentials cache name)");
+    fn++;
+    if ((r->error_code = rk_undumpdata(fn, &body, &bodylen)))
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "%s", strerror(r->error_code));
+    (void) unlink(fn);
+    free(r->ccname);
+    r->ccname = NULL;
+    if (bodylen > INT_MAX >> 4) {
+        free(body);
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "Credentials cache too large!");
+    }
+    body_b64len = rk_base64_encode(body, bodylen, &body_b64);
+    if (body_b64len == -1) {
+        free(body);
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "%s", strerror(r->error_code = errno));
+    }
+    free(body);
+    /*
+     * XXX Do better -- need to make sure to escape double quotes and control
+     * characters in `r->for_cname'!  Use heimbase.  Or libjq.
+     */
+    fprintf(r->tgts, "{\"name\":\"%s\",\"ccache\":\"%s\"}\n",
+            r->for_cname, body_b64);
+    free(body_b64);
+    return MHD_YES;
+}
+
+static heim_mhd_result
+get_tgts_param_authorize_cb(void *d,
+                            enum MHD_ValueKind kind,
+                            const char *key,
+                            const char *val)
+{
+    struct bx509_request_desc *r = d;
+    krb5_error_code ret = 0;
+
+    if (strcmp(key, "cname") != 0 || val == NULL)
+        return MHD_YES;
+
+    if (r->req == NULL) {
+        ret = hx509_request_init(r->context->hx509ctx, &r->req);
+        if (ret == 0)
+            ret = hx509_request_add_eku(r->context->hx509ctx, r->req,
+                                        ASN1_OID_ID_PKEKUOID);
+        if (ret)
+            return bad_500(r, ret, "Out of resources");
+    }
+    heim_audit_addkv((heim_svc_req_desc)r, KDC_AUDIT_VIS,
+                     "requested_krb5PrincipalName", "%s", val);
+    ret = hx509_request_add_pkinit(r->context->hx509ctx, r->req,
+                                   val);
+    if (ret)
+        return bad_403(r, ret, "Not authorized to requested TGT");
+    return MHD_YES;
+}
+
+/* For each requested principal, produce a ccache */
+static heim_mhd_result
+get_tgts_param_execute_cb(void *d,
+                          enum MHD_ValueKind kind,
+                          const char *key,
+                          const char *val)
+{
+    struct bx509_request_desc *r = d;
+    heim_mhd_result res = MHD_YES;
+
+    if (r->error_code == 0 &&
+        strcmp(key, "cname") == 0 && val) {
+        /* Handled upstairs */
+        r->for_cname = val;
+        r->error_code = k5_get_creds(r, K5_CREDS_EPHEMERAL);
+        if (r->error_code == 0)
```

Comment:
Don't fail `/get-tgts` just because we couldn't get _some_ principal's tickets.  Just note the failure in the response body JSON and move along!

Reply:
Fixed.